### PR TITLE
Port py_literal to pest with reproducibility fixes

### LINKED ISSUE #14: This crate fails to produce reproducible artifacts
I haven’t investigated too closely what exactly is it, but every time this crate is built in a controlled environment, the produced rlib is wildly different.

I suspect it might be due to `pest`, but I haven’t investigated too deeply yet.

### PR DESCRIPTION
**Pull request by nagisa**
Fixes #14.

### CODE REVIEW

File: Cargo.toml [hunk 1]
```diff
@@ -14,8 +14,10 @@ edition = "2018"
 num-bigint = { version = "0.2", default-features = false, features = ["std"] }
 num-complex = "0.2"
 num-traits = "0.2"
-pest = "1.0"
-pest_derive = "1.0"
+pest = "2.1"
+pest_derive = "2.1"
+# Pin pest_generator to version with reproducibility
+pest_generator = "2.1.2"
```
Comment (nagisa) [inline Cargo.toml]: Not released yet, hopefully they will right after/if they land https://github.com/pest-parser/pest/pull/442.

**Comment by nagisa** [comment]
Looks to be good to go.

**Comment by jturner314** [comment]
Thanks for this! I've reverted an unnecessary change in `src/parse_macros.rs`, and I've relaxed the dependency version requirements to be more broadly compatible [1]. Everything else looks good to me.

[1]: If you call `cargo update` in your project (or remove your `Cargo.lock` file), `cargo` will choose the latest compatible versions of your dependencies, including `pest_generator` 2.1.2. If you want to ensure reproducibility, you should include your `Cargo.lock` with your code to fix the versions of your dependencies. Just make sure that your `Cargo.lock` has been updated to use `pest_generator` 2.1.2.

**Comment by nagisa** [comment]
> Thanks for this! I've reverted an unnecessary change in src/parse_macros.rs

I think that change _is_ in fact necessary. Before it I would get an error of the form:

```
expected `None`, got `None`
```

because when checking it would probably call `iter.next()` (and see EOI) and then when formatting the message call `iter.next()` again to produce the nonsensical message.

**Comment by nagisa** [comment]
Either way, please do release a version that includes this sooner rather than later.

**Comment by jturner314** [comment]
> Before it I would get an error of the form:
> 
> ```
> expected `None`, got `None`
> ```

Are you referring to the unit tests (`cargo test`)? I believe the fix was when you added the `Rule::EOI` on line 96 of `parse.rs`, and I've kept that change. If I remove that, I get the error:

```
thread 'main' panicked at 'assertion failed: `(value matches pattern)`
 pattern: `Option::None`,
   value: `None`', src/parse.rs:96:24
```

I've released version 0.2.2. If you don't mind me asking, I'm just curious -- what are you using `py_literal` for?

**Comment by nagisa** [comment]
> I believe the fix was when you added the Rule::EOI on line 96 of parse.rs, and I've kept that change.

Yeah, but without the `tmp` change the error makes no sense. Though I guess I can see why you want to keep it inside the assertion...

> what are you using py_literal for?

It is a dependency of the `ndarray-npy`, which is what we are using directly (to… load npy files).

**Comment by jturner314** [comment]
Oh, I understand now what was making the error message so misleading. (As you pointed out, `debug_assert_match` was evaluating the `$value` expression twice -- once to perform the match and a second time to format the error message.) Using a `tmp` variable fixed the error message for that particular call of `debug_assert_match` because it made sure `iter.next()` was evaluated only once. I've created #16 to fix `debug_assert_match` directly. Thanks again for the PR!

**Pull request merged**